gitpod-io/gitpod
# Add basic OIDC client

This PR adds a base for OIDC clients.

What it does:
* adds `/iam/oidc/start` and `/iam/oidc/callback` handlers for OIDC auth flows.
* echos the obtained `id_token` when the OIDC auth was successful (req: Gitpod's Google Workspace account)

What's next:
* Read client configs from DB (depends on ~#15150~ #15305)
* Security relevant tests 
   * idea: `state` to be a [validated JWT](https://datatracker.ietf.org/doc/html/draft-bradley-oauth-jwt-encoded-state-09#section-3)
   * client id is matched correctly on callbacks
* call into `server` to initiate a session 



## Related Issue(s)

Part of #14955 (OIDC client implementation)
Relates to SSO Epic #7761

## How to test
There is another test deployment with a flying configuration which can be used to check this. 

Going to https://at-oidc-proxy.preview.gitpod-dev.com/iam/oidc/start will initiate the OIDC flow, which will only allow to proceed if you have a Gitpod account on Google. The result of the flow is less exciting should be an output of the id_token of yours. 

## Release Notes

```release-note
NONE
```

## Documentation


## Werft options:

      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
